internal/rootcoord/timeticksync.go

```diff
@@ -84,10 +99,63 @@ func (t *timetickSync) sendToChannel() {
 		t.proxyTimeTick[k] = nil
 	}
 	t.sendChan <- ptt
+	return nil
+}
+
+// AddDmlTimeTick add ts into ddlTimetickInfos[sourceID],
+// can be used to tell if DDL operation is in process.
+func (t *timetickSync) AddDdlTimeTick(reason string) (typeutil.Timestamp, error) {
+	t.ddlLock.Lock()
+	defer t.ddlLock.Unlock()
+
+	ts, err := t.core.TSOAllocator(1)
+	if err != nil {
+		return 0, err
+	}
+
+	if ts < t.ddlMinTs {
+		t.ddlMinTs = ts
+	}
+	t.ddlTsSet[ts] = struct{}{}
+
+	log.Debug("add ddl timetick", zap.Uint64("minTs", t.ddlMinTs), zap.Uint64("ts", ts),
+		zap.Int("len(ddlTsSet)", len(t.ddlTsSet)), zap.String("reason", reason))
+	return ts, nil
+}
+
+// RemoveDdlTimeTick is invoked in UpdateTimeTick.
+// It clears the ts generated by AddDdlTimeTick, indicates DDL operation finished.
+func (t *timetickSync) RemoveDdlTimeTick(ts typeutil.Timestamp, reason string) {
+	t.ddlLock.Lock()
+	defer t.ddlLock.Unlock()
+
+	delete(t.ddlTsSet, ts)
+	log.Debug("remove ddl timetick", zap.Uint64("ts", ts), zap.Int("len(ddlTsSet)", len(t.ddlTsSet)),
+		zap.String("reason", reason))
+	if len(t.ddlTsSet) == 0 {
+		t.ddlMinTs = typeutil.Timestamp(math.MaxUint64)
+	} else if t.ddlMinTs == ts {
+		// re-calculate minTs
+		minTs := typeutil.Timestamp(math.MaxUint64)
+		for tt := range t.ddlTsSet {
+			if tt < minTs {
+				minTs = tt
+			}
+		}
+		t.ddlMinTs = ts
+		log.Debug("update ddl minTs", zap.Any("minTs", ts))
+	}
+}
+
+func (t *timetickSync) GetDdlMinTimeTick() typeutil.Timestamp {
+	t.ddlLock.Lock()
+	defer t.ddlLock.Unlock()
+
+	return t.ddlMinTs
 }
 
 // UpdateTimeTick check msg validation and send it to local channel
-func (t *timetickSync) UpdateTimeTick(in *internalpb.ChannelTimeTickMsg) error {
+func (t *timetickSync) UpdateTimeTick(in *internalpb.ChannelTimeTickMsg, ddl bool, reason string) error {
 	t.lock.Lock()
```

Comment:
ddllock and lock is causing deadlock here.

At startTimeTickLoop function we get ddl lock.
inside update time tick we get t.lock.
if removeDdlTimeTick is called it is another ddl lock.
That's the traditional ABA deadlock

Reply:
to solve the deadlock
1) make sure lock and ddl are acquired in specific order
or 
2) remove one of the lock by using atomic operation or by syncMap

Reply:
I have move RemoveDdlTimetick outside of UpdateTimeTick, there will be no ABA deadlock.